File: pythonX/medium/m006_first_non_repeating.py
"""
M006: First Non-Repeating Character

Summary:
Find the first character that appears only once in a string.

Requirements:
Return the first non-repeating character or '\0' if none exists.

Approach:
Use dict to count frequencies, then find first char with count = 1.

Data Structures:
- dict[str, int]: character -> frequency

Big O Analysis:
Time Complexity: O(n)
Space Complexity: O(n)
"""


def first_non_repeating(s: str) -> str:
    """
    TODO: Find first non-repeating character
    1. Create dict to count frequencies
    2. First pass: count each character
    3. Second pass: find first char with count = 1
    4. Return char or '\0' if none
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic cases")
    print(f"'leetcode' -> '{first_non_repeating('leetcode')}'")      # Expected: l
    print(f"'loveleetcode' -> '{first_non_repeating('loveleetcode')}'")  # Expected: v
    print(f"'aabb' -> '{first_non_repeating('aabb')}'")              # Expected: \0
    
    print("\nTest 2: All unique")
    print(f"'abcdef' -> '{first_non_repeating('abcdef')}'")          # Expected: a
    
    print("\nTest 3: Empty string")
    print(f"'' -> '{first_non_repeating('')}'")                      # Expected: \0

File: src/com/example/javas/medium/M006_FirstNonRepeating.java
/*
MEDIUM PROBLEM (Simplified for 25 min):
Find the first non-repeating character in a string

REQUIREMENTS:
- Return the first character that appears only once
- Return '\0' (null character) if no such character exists
- Case-sensitive

APPROACH:
- Use HashMap to count frequency of each character
- Loop through string again to find first character with count = 1

DATA STRUCTURES:
- HashMap<Character, Integer>: character -> frequency count

TIME COMPLEXITY: O(n)
SPACE COMPLEXITY: O(n)
*/

package com.example.javas.medium;

import java.util.*;

public class M006_FirstNonRepeating {
    
    public static char firstNonRepeating(String s) {
        // TODO: Implement first non-repeating character finder
        // 1. Create HashMap to store character frequencies
        // 2. First pass: count frequency of each character
        // 3. Second pass: find first character with frequency = 1
        // 4. Return that character, or '\0' if none found
        
        return '\0';
    }
    
    // TEST CASES
    public static void main(String[] args) {
        System.out.println("\"leetcode\" -> '" + firstNonRepeating("leetcode") + "'");
        System.out.println("\"loveleetcode\" -> '" + firstNonRepeating("loveleetcode") + "'");
        System.out.println("\"aabb\" -> '" + firstNonRepeating("aabb") + "'");
        System.out.println("\"abcdef\" -> '" + firstNonRepeating("abcdef") + "'");
        System.out.println("\"\" -> '" + firstNonRepeating("") + "'");
    }
}
